### PR TITLE
ENH: stats: add Baumgartner-Weiss-Schindler test

### PR DESCRIPTION
**Pull request by zhimin-z**
[skip ci]

#### Reference issue
Create BWS test for statistical purposes (both naive and variant [Shan + Murakami] methods) (both one-sided and two-sided)

#### What does this implement/fix?
The Baumgartner-Weiss-Schindler (BWS) test is a nonparametric test of the null hypothesis that the distribution underlying sample `x` is the same as the distribution underlying sample `y`. Unlike the Kolmogorv-Smirnov, Wilcoxon, and Cramer-Von Mises tests, the Baumgartner-Weiss-Schindler test weights the integral by the variance of the difference in CDFs, emphasizing the tails of the distributions, which increases the power of the test for a lot of applications.

#### Additional information
Continue on https://github.com/scipy/scipy/pull/18317

### CODE REVIEW

**Comment by mdhaber** [comment]
@chrisb83 requested more information [here](https://mail.python.org/archives/list/scipy-dev@python.org/thread/NBWPQPVGMMOSN7JI6434IAB4OOPHZFMD/). I'd suggest answering his question.

Otherwise, I would prefer not to write a task list until it has been decided whether this test should be included in SciPy. So far, I haven't seen any explicit support. 

Also, please take a look at the conversation beginning at https://github.com/scipy/scipy/pull/18317#issuecomment-1520727689. If you are using a bot, please disable it so that this doesn't happen again.

**Comment by zhimin-z** [comment]
> @chrisb83 requested more information [here](https://mail.python.org/archives/list/scipy-dev@python.org/thread/NBWPQPVGMMOSN7JI6434IAB4OOPHZFMD/). I'd suggest answering his question.
> 
> Also, please take a look at the conversation beginning at [#18317 (comment)](https://github.com/scipy/scipy/pull/18317#issuecomment-1520727689). If you are using a bot, please disable it so that this doesn't happen again. If we move forward with this, please do not force push at all.
> 
> Otherwise, I would prefer not to write a task list until it has been decided whether this test should be included in SciPy. So far, I haven't seen any explicit support.

Thanks, but as I mentioned earlier in [#18317](https://github.com/scipy/scipy/pull/18317) , bws is commonly used in academia, my recent paper on mining software repository used bws test as well. @mdhaber

**Comment by mdhaber** [comment]
Thanks.
@chrisb83, the answer to your question is above.

**Comment by chrisb83** [comment]
> Thanks. @chrisb83, the answer to your question is above.

Thanks, the papers above don't appear very relevant (e.g., comparing only against KS or Wilcoxon based in the abstracts). But given that there is interest in the test + the original paper reports advantages compared to CVM, there are enough arguments to add it.

**Comment by zhimin-z** [comment]
> @chrisb83 requested more information [here](https://mail.python.org/archives/list/scipy-dev@python.org/thread/NBWPQPVGMMOSN7JI6434IAB4OOPHZFMD/). I'd suggest answering his question.
> 
> Also, please take a look at the conversation beginning at [#18317 (comment)](https://github.com/scipy/scipy/pull/18317#issuecomment-1520727689). If you are using a bot, please disable it so that this doesn't happen again. If we move forward with this, please do not force push at all.
> 
> Otherwise, I would prefer not to write a task list until it has been decided whether this test should be included in SciPy. So far, I haven't seen any explicit support.

Based on the conversation, maybe we can make a task list for this test. This is my first time creating PR on Scipy, wondering how a PR is qualified for a merge. @chrisb83 @mdhaber

**Comment by zhimin-z** [comment]
> For the "protocol", I am -0.5 on adding this test if there are no strong advantages (the authors paper saying it's better is not very convincing to me. OC it's better somehow to publish 😅). Citation counts are not everything, but seems a bit on the low side as well. I am still wondering how this method is really influential and fundamental. And if we do add it (not blocking anything just to be clear), we need to be clear in the doc about why use this and when.
> 
> All that being said 😃 @zhimin-z you can find a lot of information in our contributing guides: https://scipy.github.io/devdocs/dev/contributor/contributor_toc.html. And this page is a quick summary with a sort of checklist at the end https://scipy.github.io/devdocs/dev/contributor/development_workflow.html But we don't work with checklist so you can keep this on your side if you want but it will not "help" or do much for the review itself.

For the reason why bws test is better, I would refer to the results from this repo: https://github.com/cran/BWStest. I understand the citation count is low partly due to the permutation test efficiency being low compared to other rank-based tests, but its comparative power is higher when we are examining small datasets.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:93]:
```suggestion
    the Kolmogorov-Smirnov, Wilcoxon, and Cramer-Von Mises tests, 
```

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:84]: Please see https://scipy.github.io/devdocs/dev/missing-bits.html#return-objects about return objects.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:87]: Please make arguments after `y` keyword only.
 https://scipy.github.io/devdocs/dev/missing-bits.html#required-keyword-names

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:103]: It's fine to leave `axis` in the code, but I'd recommend removing the public argument *for now*. When we add `_axis_nan_policy` decorator in a follow-up, we'll expose it. The advantage is that you won't have to write tests.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:128]: I don't think there should be duplicates. Consider `shan-murakami`.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:129]: This is fine for now, but it will be replaced by the `PermutationMethod` class of gh-18227 before this merges.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:121]: Something I asked in the original PR - is having these variants worth the extra complexity?
Only one person has asked for this test in the past few decades, so I wonder whether the variant would be missed. 

If the "naive" option is naive and missing something that murakami accounts for, why not only use that?

What are your thoughts?

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:157]:
```suggestion

    See Also
```

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:164]: Missing newline, I think

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:183]: Identifications of what?

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:205]:
```suggestion
    by `bws_test` also agrees with :math:`p = 0.0029` reported in [1]_.
```

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:210]: The default number if resamples is 9999. If an exact test can be performed with fewer samples, it will be. Otherwise, a randomized test is performed.

**Review by mdhaber — DISMISSED** [review]
Please add unit tests:
- compare the output against external references
- check that all parameters do the right thing
- check that invalid arguments produce the desired error messages
- check for correct behavior with edge cases (e.g. empty `x` or `y`)

Since there is no asymptotic distribution here, I'm considering whether this should be generalized into an `equivalence_test` with several statistic options, just as `goodness_of_fit` unifies the one-sample goodness of fit tests. Please don't do that here, but what are your thoughts?

File: scipy/stats/_bws_test.py
Comment (zhimin-z) [reply scipy/stats/_bws_test.py]: yeah, it might be better

File: scipy/stats/_bws_test.py
Comment (zhimin-z) [reply scipy/stats/_bws_test.py:124]: All fixed, any other review suggestions?

**Comment by mdhaber** [comment]
Thanks for these updates. There are two unresolved comments:
- [ ] Add tests (specific recommendations [here](https://github.com/scipy/scipy/pull/18363#pullrequestreview-1404206322))
- [ ] Justify the need for both variants or offer only one variant ([here](https://github.com/scipy/scipy/pull/18363#discussion_r1179262555))

After that is done, I may suggest a few tweaks to the documentation or refactoring with a PR to your branch, but I think it's pretty close.

**Comment by zhimin-z** [comment]
> Thanks for these updates. There are two unresolved comments:
> 
> * [ ]  Add tests (specific recommendations [here](https://github.com/scipy/scipy/pull/18363#pullrequestreview-1404206322))
> * [ ]  Justify the need for both variants or offer only one variant ([here](https://github.com/scipy/scipy/pull/18363#discussion_r1179262555))
> 
> After that is done, I may suggest a few tweaks to the documentation or refactoring with a PR to your branch, but I think it's pretty close.

- I read the original paper, but it did not offer me a reproducible validation test, instead, they use the permutation sampling test to compare the power of each k-sample test. I wonder how and where I can find data to add tests for variant methods.
- The variants method now combines into one single option name `shan-murakami`.

**Comment by mdhaber** [comment]
> I read the original paper, but it did not offer me a reproducible validation test, instead, they use the permutation sampling test to compare the power of each k-sample test. I wonder how and where I can find data to add tests for variant methods.

You can compare the output of this implementation against an [R package](https://www.google.com/search?q=baumgartner+weiss+schindler+r&rlz=1C1ONGR_enUS930US930&oq=baumgartner+weiss+schindler+r&aqs=chrome..69i57j35i39j0i390i650l3j69i61l3.2879j0j7&sourceid=chrome&ie=UTF-8). (Just don't look at the source code.)

> The variants method now combines into one single option named shan-murakami.

Yes, I saw. (And I saw that the line `variants = {"shan", "murakami", "naive"}` was not updated accordingly.) Please see the comment linked from the checklist. The question is why not offer *only* the "naive" method or *only* the *Shan-Murakami* method and eliminate the `variant` keyword? Why is it so important to include both?

**Comment by zhimin-z** [comment]
> 

For the variant methods, I got the inspiration from [Kendall Tau rank test](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.kendalltau.html)
image
Do you think this is a good reason to go?

**Comment by mdhaber** [comment]
The fact that `kendalltau` has two variants is not sufficient support for this test to be added with two variants. 

We have variants of `kendalltau` because it is a very popular function and users requested that other variants be added.

**Comment by zhimin-z** [comment]
> The fact that `kendalltau` has two variants is not sufficient support for this test to be added with two variants.
> 
> We have variants of `kendalltau` because it is a very popular function and users requested that other variants be added.
> 
> _If_ there were a reason for this function to get multiple variants, `variant` would be a good name for the keyword, though.

I've checked the BWS test in R and found that it has the variant "murakami" in their package.

I also checked the kendalltau test and found the variants has only two options.

What do you suggest to change the name for?

**Comment by mdhaber** [comment]
@zhimin-z it sounds like we're misunderstanding one another. Rather than clarify the conversation above, let's start over and pretend that this is a totally new comment that we haven't discussed before.

---

What is the advantage of "Shan-Murakami" over "Naiive" (or vice-versa)?

**Comment by zhimin-z** [comment]
> 
Extracted from [original paper](https://www.jstage.jst.go.jp/article/jjscs1988/19/1/19_1_1/_pdf) page 7:

and another paper page 13:

**Comment by zhimin-z** [comment]
I would add more variants as shown in https://www.tandfonline.com/doi/pdf/10.1080/00949655.2010.551516
image

**Comment by mdhaber** [comment]
I see you added more variants. This is actually what I was trying to prevent. Once you add a `variant` keyword, it's easy to find more variants that *could* be added. This complicates the API, review, and future maintenance. (It's hard to argue that there is a proportional benefit to users when only one person has ever asked for the test, let alone all the possible variants.)

I'd suggest focusing on adding the unit tests.

**Comment by mdhaber** [comment]
Please consider accepting zhimin-z/scipy#18. After that and a final pass, I think I would be comfortable merging the PR.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [inline scipy/stats/_bws_test.py:1]:
```suggestion
```

**Comment by mdhaber** [comment]
@chrisb83 I trimmed down the options and added unit tests so we can get this merged. I think you were the only maintainer to voice interest in adding this, so would you be willing to review it?

**Comment by mdhaber** [comment]
@chrisb83 this PR (almost literally) has your name on it. Care to provide a second review? I helped it along because you supported the addition of the function. But because I worked on it a bit, I think it could use an independent set of eyes.

**Comment by chrisb83** [comment]
Ok, I can take a look in the next few days!

File: scipy/stats/_bws_test.py
Comment (chrisb83) [inline scipy/stats/_bws_test.py:39]: I find this line hard to read, could this be re-written as 
```
if ...:
   ...
else:
   ...
```

File: scipy/stats/_bws_test.py
Comment (chrisb83) [inline scipy/stats/_bws_test.py:38]: remove outer ()

File: scipy/stats/_bws_test.py
Comment (chrisb83) [inline scipy/stats/_bws_test.py:61]: minor point:  cumulative distribution functions (CDFs)

**Review by chrisb83 — COMMENTED** [review]
I just noted a few minor things.

Nice to see how easy it is to implement such a test if `permutation_test` is used

File: scipy/stats/_bws_test.py
Comment (mdhaber) [reply scipy/stats/_bws_test.py:39]: Refactored for readability.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [reply scipy/stats/_bws_test.py:34]: Thanks for the reminder. For simplicity in their first PR, I asked the author to remove the `axis` argument from the public function. However, when `permutation_test` sees the `axis` argument in the statistic function, it calls it in a vectorized way for efficiency. Added `axis` argument, so it's much faster now.

File: scipy/stats/_bws_test.py
Comment (mdhaber) [reply scipy/stats/_bws_test.py:94]: Artifact of previous version. Fixed.

File: scipy/stats/tests/test_hypotests.py
Comment (mdhaber) [reply scipy/stats/tests/test_hypotests.py:1806]: > no need to change it

Great; I'll let it be then. I've done a fair amount of work on scipy.stats and I land on the other side here. If you'd like, I'd be happy to discuss it.

**Comment by mdhaber** [comment]
Thanks @chrisb83. I think I addressed all comments.

**Comment by chrisb83** [comment]
> Thanks @chrisb83. I think I addressed all comments.

Yes, it looks good now. Test cases against R and the paper are implemented, I don't have other feedback.

**Comment by mdhaber** [comment]
Thanks @chrisb83. If you're confident enough to merge, great, or if not, how about the approve button?

**Review by chrisb83 — APPROVED** [review]
> Thanks @chrisb83. If you're confident enough to merge, great, or if not, how about the approve button?

I am happy to approve. The status says that you requested changes, could you change that before I merge just to be sure? Thx

**Review by mdhaber — COMMENTED** [review]
The statistic magnitude agrees closely with that produced by BWSTest in R. For consistency with the rest of `scipy.stats`, we decided that (unlike BWSTest) we would _not_ modify the sign of the returned statistic depending on whether the alternative hypothesis is 'less' or 'greater'. Consequently, the relationship between the data, the statistic, the p-value, and the alternative is natural (e.g. when the distribution underlying `x` is _greater_ than the distribution underlying `y`, the statistic will be _positive_, and the p-value with `alternative='greater'` will be small).

I cannot find extensive reference values for the exact p-value, but:
- Our p-value is produced by `permutation_test`, which is thoroughly tested
- Our p-value agreed exactly with the one exact reference p-value we have (from [1], `10/3432`)
- Our p-value tends to agree few digits to that produced by BWSTest in R
- Repetitions of the test with data produced under the null hypothesis results in p-values that appear to be uniformly distributed.

<details>

```python3
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
rng = np.random.default_rng(1638083107694713882823079058616272161)

ps = []

for i in range(10000):
    print(i)
    x = rng.random(size=6)
    y = rng.random(size=3)
    res = stats.bws_test(x, y, alternative='less')
    ps.append(res.pvalue)

plt.hist(ps)
plt.show()

stats.probplot(ps, (0, 1), dist='uniform', plot=plt)
plt.axis('equal')
plt.show()
```

</details> 
Quality of documentation and tests is consistent with other stats functions.

I'll let @chrisb83 do the honors, since I was involved with the implementation. 

I thought I'd mention that I would prefer that we do not add other variants of this test.

**Review by tupui — APPROVED** [review]
Thank you all, this is in a good state. The API looks sensible the tests look good, the doc too and this was reviewed by both Christoph and Matt. Let's get this in 😃 

Thank you @zhimin-z for your first PR and working with us 🚀

**Pull request merged**
